<div style="width: 100%; clear: both;">
<div style="float: left; width: 50%;">
<img src="http://www.uoc.edu/portal/_resources/common/imatges/marca_UOC/UOC_Masterbrand.jpg", align="left">
</div>
<div style="float: right; width: 50%;">
<p style="margin: 0; padding-top: 22px; text-align:right;">M1.304 · Inteligencia Artificial Avanzada / M0.539 · Inteligencia Artificial</p>
<p style="margin: 0; text-align:right;">MU Ingeniería Informática / MU Ingeniería Computacional y Matemática</p>
<p style="margin: 0; text-align:right; padding-button: 100px;">Estudios de Informática, Multimedia y Telecomunicación</p>
</div>
</div>
<div style="width:100%;">&nbsp;</div>

## PEC 1: SISTEMAS RECOMENDADORES



### Presentación

En esta PEC trabajaréis con sistemas recomendadores, tanto a bajo nivel (programando funciones y matrices de similitud) como a un nivel más alto (mediante la biblioteca Surprise).

### Competencias

En este enunciado se trabajan las siguientes competencias generales de
máster:
- Capacidad para proyectar, calcular y diseñar productos, procesos e
instalaciones en todos los ámbitos de la ingeniería informática.
- Capacidad para el modelado matemático, cálculo y simulación en
centros tecnológicos y de ingeniería de empresa, particularmente en
tareas de investigación, desarrollo e innovación en todos los ámbitos
relacionados con la ingeniería informática.
- Capacidad para aplicar los conocimientos adquiridos y solucionar
problemas en entornos nuevos o poco conocidos dentro de contextos
más amplios y multidisciplinares, siendo capaces de integrar estos
conocimientos.
- Poseer habilidades para el aprendizaje continuo, autodirigido y
autónomo.
- Capacidad para modelar, diseñar, definir la arquitectura, implantar,
gestionar, operar, administrar y mantener aplicaciones, redes,
sistemas, servicios y contenidos informáticos.

Las competencias específicas de esta asignatura que se trabajan en esta
prueba son:
- Entender qué es el aprendizaje automático en el contexto de la
inteligencia artificial.
- Distinguir entre los diferentes tipos y métodos de aprendizaje.
- Aplicar las técnicas estudiadas a un caso real.

### Objectivos

En esta PEC aprenderéis como funcionan los sistemas recomendadores. Más concretamente, aprenderéis a trabajar con funciones de similitud y a hacer predicciones con una biblioteca de uso habitual en el área: Surprise.

### Recursos

Esta PEC requiere los recursos siguientes:

Archivos proporcionados:

  * El archivo .ipynb que estáis leyendo ahora mismo y que deberéis completar para resolver la PEC.
  * El archivo DATASET.CSV, el cual contiene los datos con los que vais a trabajar.

Complementarios: 
  * Manual de teoría de la asignatura.
  * Aocumentación de las librerías utilizadas, entre las que destacan Pandas, NumPy y surprise.

### Entrega y criterios de evaluación

La PEC se debe entregar el **29 de marzo del 2022**.

La entrega debe incluir una versión editada de este cuaderno (.ipynb). Se recomienda el uso de Google Colab (https://colab.research.google.com/). El código de las soluciones a los ejercicios se debe implementar y ejecutar en las celdas de código proporcionadas y la respuestas justificadas se deben agregar a las celdas de texto correspondientes.

Todas las respuestas deben estar correctamente razonadas y justificadas. **Las soluciones que no vayan acompañadas de la correspondiente respuesta razonada no serán evaluadas**.

Junto a cada una de las actividades se indica su puntuación. La evaluación de cada apartado considerará la respuesta proporcionada, la corrección y la optimización del código y también la claridad, tanto del código como de las explicaciones que se soliciten.

### Descripción de la PEC

#### Descripción de los datos

Los datos con los que trabajaréis provienen del "User-Book Interactions dataset". Se trata de un conjunto de datos (o *dataset*) que recoge valoraciones de libros hechas por distintos usuarios. Encontraréis información detallada en su sitio web:

https://sites.google.com/eng.ucsd.edu/ucsdbookgraph/shelves?authuser=0

Este conjunto de datos se ha utilizado en las siguientes publicaciones:

* Mengting Wan, Julian McAuley, *Item Recommendation on Monotonic Behavior Chains*, in RecSys'18.

* Mengting Wan, Rishabh Misra, Ndapa Nakashole, Julian McAuley, *Fine-Grained Spoiler Detection from Large-Scale Review Corpora*, in ACL'19.

Sin embargo no debéis trabajar con el dataset antrior sino con el que os proporcionamos (DATASET.CSV), el cual es una versión reducida del original.

El formato de DATASET.CSV es simple.

* La primera fila es la cabecera y contiene los nombres de las tres columnas: user_id, book_id, rating.
* Las siguientes filas contienen los datos. Cada fila se corresponde a una valoración de un libro hecha por un usuario. En particular, en cada fila encontraréis el identificador del usuario que ha hecho la valoración (user_id), el identificador del libro que ha valorado (book_id) y la valoración hecha (rating).

Las valoraciones son números entre 0 (el libro no ha gustado) y 5 (el libro ha gustado mucho). Los identificadores son simplemente números que identifican unívocamente cada usuario y cada libro.

#### Preparación del entorno

Para realizar esta PEC deberéis programar en Python y utilizar algunas de sus bibliotecas más conocidas, como Pandas o NumPy. Estas dos bibliotecas se encuentran preinstaladas en la mayoría de distribuciones de Python. Si no es vuestro caso, podéis instalarlas ejecutando

* !pip install pandas
* !pip install numpy

Podéis encontrar información detallada sobre estas dos bibliotecas en:

* https://pandas.pydata.org/
* https://numpy.org/

También utilizaréis la biblioteca surprise:

* http://surpriselib.com/

Esta biblioteca no se encuentra preinstalada en muchas distribuciones de Python. Más adelante se os indicará como instalarla.

La entrega de la PEC consistirá en esta plantilla ipynb debidamente completada, tal y como se indica más adelante. Para ello podéis trabajar en local mediante Jupyter Lab o similares o en remoto mediante Google Colab. Nuestra recomendación es que utilicéis Google Colab, accediendo a él mediante el usuario de Google de vuestra cuenta de la UOC, ya que las bibliotecas más relevantes se encuentran preinstaladas en Colab. Algunas, no obstante, no lo están. En particular, Pandas y NumPy ya están preinstaladas en Colab (no es necesario ejecutar las líneas anteriores) mientras que Surprise no lo está (deberéis instalarla como se indicará más adelante).

A continuación se proporcionan instrucciones para preparar el entorno de trabajo tanto si es local como si es remoto (Google Colab).

Si trabajáis en local (Jupyter Lab o similares) debéis ubicar el archivo DATASET.CSV en la misma carpeta donde se encuentre este ipynb y asignar True a la variable *localDataset*.

Si trabajáis en Google Colab deberéis ubicar el archivo DATASET.CSV en el Google Drive de la misma cuenta donde ejecutáis Colab. Lo más recomendable es que lo hagáis todo des de la cuenta de Google de la UOC. Una vez hecho este deberéis asignar False a la variable *localDataset* y modificar la variable *remotePath* (también en la siguiente celda) para que apunte a la carpeta de Google Drive donde habéis guardado DATASET.CSV.

En cualquiera de los dos casos debéis ejecutar la siguiente celda tras seguir las instrucciones anteriores. También en los dos casos, la siguiente celda almacenará la ruta a DATASET.CSV dentro de la variable *localFileName*. Por lo tanto, cuando tengáis que cargar el dataset deberéis hacerlo utilizando *localFileName* como nombre de archivo.

Notad que si trabajáis en Colab el script de la siguiente celda monta vuestra unidad de Google Drive dentro del espacio de Colab. Por este motivo se os solicitará que autoricéis a Colab el acceso a vuestro Google Drive.

Notad también que deberéis ejecutar la siguiente celda cada vez que empecéis a trabajar en este ipynb, tanto si trabajáis en local como si lo hacéis en Colab. El propio archivo ipynb se guarda automáticamente en vuestro espacio de Google y podéis también guardarlo en local para hacer la entrega de la PEC. Ahora bien, los archivos que guardéis en vuestro espacio Colab **se pierden al cerrar la sesión**.

La realización de la PEC no requiere que se guarde ninún archivo. Ahora bien, si por algún motivo lo hacéis recordad que para conservarlo deberéis copiarlo desde Colab a vuestro Drive (o bajarlo a vuestro ordenador) y, probablemente, copiarlo de nuevo a vuestro espacio Colab al volver a acceder a él. 

Encontraréis información detallada sobre estos aspectos en el siguiente link:

https://colab.research.google.com/notebooks/io.ipynb


In [ ]:
! pip install pandas numpy

In [1]:
# FOLLOW THE PREVIOUS INSTRUCTIONS AND RUN THE CELL EVERY TIME YOU START WORKING

localDataset=True

if not localDataset:
    from google.colab import drive
    import shutil
    import os
    mountPoint='/content/drive'
    remotePath='MyDrive/PAC1_IAA_2122'
    localPath='.'
    fileName='DATASET.CSV'
    remoteFileName=os.path.join(mountPoint,remotePath,fileName)
    localFileName=os.path.join(localPath,fileName)

    print('[ COPYING DATA FROM GOOGLE DRIVE TO LOCAL COLAB SPACE ]')
    if os.path.isfile(localFileName):
      print('  * FILE %s ALREADY PRESENT.'%localFileName)
    else:
      if os.path.isdir(mountPoint):
        print('  * GOOGLE DRIVE ALREADY MOUNTED AT %s'%mountPoint)
      else:
        print('  * MOUNTING GOOGLE DRIVE AT %s'%mountPoint)
        drive.mount(mountPoint)
      print('  * COPYING FROM %s TO %s'%(remoteFileName,localFileName))
      shutil.copy(remoteFileName,localFileName)
else:
    print('[ WORKING LOCALLY ]')
    localFileName='DATASET.CSV'
    
print('[ DONE ]')

[ WORKING LOCALLY ]
[ DONE ]


### Ejercicios

Cada Actividad es independiente de las otras. Por lo tanto, si bien es recomendable realizarlas en el orden propuesto, es posible resolverlas en cualquier orden.

Notad que el comentario anterior se refiere a las Actividades. Los apartados dentro de cada Actividad sí deben realizarse en el orden propuesto.

#### **Actividad 1: Comprender los datos**

Para realizar esta actividad utilizaréis bibliotecas como Pandas y NumPy. Por lo tanto, es importante que os documentéis sobre su funcionamiento. En particular, Pandas permite filtrar o pivotar datos cómodamente y NumPy permite realizar de forma eficiente cálculos y búsquedas de datos en matrices numéricas. En este sentido os pueden ser útiles funciones de NumPy como argmax y argmin. Tened en cuenta también que es posible convertir datos desde un dataframe de Pandas a una matriz NumPy mediante las funciones existentes en dichas bibliotecas.

**a) (0.25 puntos)** Cargad el dataset como un dataframe de Pandas denominado dataSet. Aunque en futuras acctividades podéis crear (si lo necesitáis por algú motivo) versiones alternativas, filtradas, modificadas, procesadas, ... del dataset, la variable dataSet no debe modificarse y siempre debe contener los datos tal y como se carguen en este apartado.

In [2]:
import pandas as pd
import numpy as np

dataSet = pd.read_csv("DATASET.CSV")

**b) (0.75 puntos)** Completa la siguiente celda de manera que se ponga dentro de las variables indicadas la información que se lista a continuación. El tipo de las variables debe coincidir con el del formato utilizado en los "print" de forma que los resuldatos se visualicen correctamente.

* numEntries debe contener el número de registros (filas) del dataset.
* numAttributes debe contener el número de atributos (columnas) del dataset.
* numUsers debe contener el número de usuarios distintos que aparecen en el dataset.
* numBooks debe contener el número de libros distintos que se han valorado en el dataset.
* meanRate debe contener la valoración media del dataset.
* exampleRate debe contener la valoración hecha por el usuario con identificador 361926 del libro con identificador 301,

In [3]:
# COMPLETE AND EXECUTE THE CELL

numEntries=len(dataSet)
numAttributes=len(dataSet.columns)
numUsers=len(pd.unique(dataSet['user_id']))
numBooks=len(pd.unique(dataSet['book_id']))
meanRate=dataSet['rating'].mean()
exampleRate=dataSet.loc[(dataSet['user_id'] == 361926) & (dataSet['book_id'] == 301), 'rating'].iat[0]


# DO NOT MODIFY THE FOLLOWING LINES

print('ENTRIES         : %d'%numEntries)
print('ATTRIBUTES      : %d'%numAttributes)
print('DIFFERENT USERS : %d'%numUsers)
print('DIFFERENT BOOKS : %d'%numBooks)
print('AVERAGE RATING  : %.3f'%meanRate)
print('EXAMPLE RATING  : %d'%exampleRate)

ENTRIES         : 9835
ATTRIBUTES      : 3
DIFFERENT USERS : 2955
DIFFERENT BOOKS : 513
AVERAGE RATING  : 3.844
EXAMPLE RATING  : 5


**c) (0.75 puntos)** Completa la siguiente celda de manera que cada una de las variables indicadas contenga el valor que se indica a continuación. El tipo de cada variable debe coincidir con el del formato utilizado en los "print" de forma que los resultados se visualicen correctamente. Dentro de la celda de la respuesta podéis incluir funciones o código auxiliar si lo necesitáis. Si hay varios valores posibles para una variable, asignadle uno de ellos.

* mostProlificUser debe contener el identificador del usuario que más libros ha valorado.
* maxVotes debe contener el número de libros que ha valorado el usuario mostProlificUser.
* leastProlificUser debe contener el identificador del usuario que menos libros ha valorado.
* minVotes debe contener el número de libros que ha valorado el usuario leastProlificUser.
* meanVotes debe contener el número promedio de valoraciones hechas por los usuarios.

In [4]:
# COMPLETE AND EXECUTE THE CELL

mostProlificUser=dataSet['user_id'].mode().iat[0]
maxVotes=dataSet.loc[dataSet['user_id'] == mostProlificUser].count()['book_id']
leastProlificUser=dataSet['user_id'].value_counts().index[-1]
minVotes=dataSet.loc[dataSet['user_id'] == leastProlificUser].count()['book_id']
meanVotes=dataSet.user_id.map(dataSet.user_id.value_counts()).mean()

# DO NOT MODIFY THE FOLLOWING LINES

print('MOST PROLIFIC USER      : %d'%mostProlificUser)
print('MAX. NUMBER OF VOTES    : %d'%maxVotes)
print('LEAST PROLIFIC USER     : %d'%leastProlificUser)
print('MIN. NUMBER OF VOTES    : %d'%minVotes)
print('AVERAGE NUMBER OF VOTES : %.3f'%meanVotes)

MOST PROLIFIC USER      : 88757
MAX. NUMBER OF VOTES    : 18
LEAST PROLIFIC USER     : 38186
MIN. NUMBER OF VOTES    : 1
AVERAGE NUMBER OF VOTES : 5.427


**d) (0.75 puntos)** Completa la siguiente celda de manera que cada una de las variables indicadas contenga el valor que se indica a continuación. El tipo de cada variable debe coincidir con el del formato utilizado en los "print" de forma que los resultados se visualicen correctamente. Dentro de la celda de la respuesta podéis incluir funciones o código auxiliar si lo necesitáis. Si hay varios valores posibles para una variable, asignadle uno de ellos.

* mostReadBook debe contener el identificador del libro que más usuarios han valorado.
* maxUsers debe contener el número de usuarios que han valorado el libro mostReadBook.
* leastReadBook debe contener el identificador del libro que menos usuarios han valorado.
* minUsers debe contener el número de usuarios que han valorado el libro leastReadBook.
* meanUsers debe contener el número medio de usuarios que han valorado cada libro.

In [5]:
# COMPLETE THE CELL

mostReadBook=dataSet.book_id.mode().iat[0]
maxUsers=dataSet.loc[dataSet['book_id'] == mostReadBook]['user_id'].count()
leastReadBook=dataSet.book_id.value_counts().index[-1]
minUsers=dataSet.loc[dataSet['book_id'] == leastReadBook]['user_id'].count()
meanUsers=(maxUsers+minUsers)/2

# DO NOT MODIFY THE FOLLOWING LINES

print('MOST READ BOOK          : %d'%mostReadBook)
print('MAX. USERS              : %d'%maxUsers)
print('LEAST READ BOOK         : %d'%leastReadBook)
print('MIN. USERS              : %d'%minUsers)
print('AVERAGE NUMBER OF USERS : %.3f'%meanUsers)

MOST READ BOOK          : 536
MAX. USERS              : 85
LEAST READ BOOK         : 23819
MIN. USERS              : 1
AVERAGE NUMBER OF USERS : 43.000


**e) (0.75 puntos)** Sabiendo que queremos construir un recomendador de libros y basándoos en los resultados de los apartados anteriores, creéis que la distribución de los datos es adecuada? Justificad la respuesta.

Para construir un recomendador de libros basándonos en técnicas de _Machine Learning_ la distribución de datos es adecuada ya que está dividida en 3 columnas "user_id", "book_id" y "rating" y este es el formato esperado en algunas librerías como surprise. Sin embargo, podrían existir más puntuaciones de los usuarios frente a muchos más libros, teniendo en cuenta el volumen del dataset vemos que el usuario con mayor número de votos ha emitido solo 18 y que la media de votos es de 5, muy poca dado el volumen de libros.

#### **Actividad 2: Similitudes**

**a) (1.25 puntos)** Programad la función *compute_similarity* que recibe como parámetros de entrada dos identificadores de usuario (*firstUser* y *secondUser*) y retorna la similitud Euclídea entre los dos (*theSimilarity*) y también el número de libros que han valorado en común los dos usuarios (*numCommon*). Para calcular la similitud Euclidea debéis:

* Utilizar sólo  los libros valorados por los dos usuarios.
* Calcularla a partir de la distancia Euclidea como s=1/(1+d), donde s es la similitud y d es la distancia Euclidea.

Para programar la función tan solo podéis utilizar las bibliotecas Pandas y NumPy. La función puede suponer que *dataSet* es una variable global. Además, si debéis precalcular algún dato para acelerar los cálculos, hacedlo también en una variable global.

Utilizad la definición que ya se os proporciona, manteniendo el "return" como su última línea. El código que se proporciona a continuación de la función no debe modificarse y debe mostrar los resultados correctamente al ejecutar la celda.

In [6]:
import numpy as np

def compute_similarity(firstUser,secondUser):
    df = pd.merge(dataSet.loc[dataSet.user_id == firstUser], dataSet.loc[dataSet.user_id == secondUser], on='book_id')    
    dist = np.sqrt(np.sum(np.square(np.array(df['rating_x']) - np.array(df['rating_y']))))
    simil = 1 / (1 + dist)
    return simil, len(df['rating_x'])

# DO NOT MODIFY THE FOLLOWING LINES

# As an example, compute the similarity and the number of common books
# between users 19948 and 433738
firstUser=19948
secondUser=433738

theSimilarity,numCommon=compute_similarity(firstUser,secondUser)
print('Users %d and %d rated %d books in common, resulting in a similarity of %.3f'%(firstUser,secondUser,numCommon,theSimilarity))

Users 19948 and 433738 rated 4 books in common, resulting in a similarity of 0.205


**b) (1.25 puntos)** Programad la función *compute_similarity_matrix* la cual debe construir dos matrices:

* *similarityMatrix* debe ser la matriz de simitudes Euclideas. La matriz debe contener todas las similitudes Euclideas entre todas las posibles parejas de usuarios de forma que similarityMatrix[r,c] debe contener la similitud Euclidea entre el usuario con identificador allUsers[r] y el usuario con identificador allUsers[c]. Aunque, por definición, similarityMatrix[x,x]=1 para todos los posibles valores de x, asignad un 0 a estas posiciones (la diagonal) para prevenir problemas en actividades posteriores.
* *commonMatrix* debe contener en cada celda [r,c] el número de libros en común que han valorado los usuarios con identificador allUsers[r] y allUsers[c].

Ambas matrices serán cuadradas y simétricas y tendrán tantas filas (y columnas) como usuarios distintos existan en el dataset. Ya se or proporciona la línea de código que costruye el vector allUsers. La función no tiene parámetros de entrada pero puede utilizar funciones y datos globales ya definidos o calculados anteriormente. Como en actividades anteriores, respetad la cabecera de la función y el "return" proporcionados.

Notad que el tiempo que tarde en ejecutarse la función puede depender en gran medida de la optimización de funciones anteriores. Si el código (de esta función y de las anteriores) está poco optimizado, el tiempo de ejecución puede ser muy alto.

In [ ]:
# COMPLETE AND EXECUTE THE CELL

import numpy as np

# Get all the user ids
allUsers=np.array(dataSet['user_id'].unique())


# The requested function
def compute_similarity_matrix():
    similarityMatrix = []
    commonMatrix = []
    for user_1 in allUsers:
        simils = []
        commons = []
        for user_2 in allUsers:
            if user_1 == user_2:
                simil = 0
            else:
                simil, common = compute_similarity(user_1, user_2)
                commons.append(common)
            simils.append(simil)
        similarityMatrix.append(simils)
        commonMatrix.append(commons)
    return similarityMatrix,commonMatrix

# DO NOT MODIFY THE FOLLOWING LINES

similarityMatrix,commonMatrix=compute_similarity_matrix()

# As an example, get the similarity and common reviews between users 19948 and 433738
firstUser=19948
secondUser=433738
idxFirst=int(np.argwhere(allUsers==firstUser))
idxSecond=int(np.argwhere(allUsers==secondUser))
theSimilarity=similarityMatrix[idxFirst,idxSecond]
numCommon=commonMatrix[idxFirst,idxSecond]
print('Users %d and %d rated %d books in common, resulting in a similarity of %.3f'%(firstUser,secondUser,numCommon,theSimilarity))

**c) (1.25 puntos)** Utilizando la matriz *similarityMatrix* construida en la celda anterior (y que podéis considerar una variable global), programad la función *recommend_books*. Esta función debe recibir como parámetro de entrada un identificador de usuario (*theUser*) y debe retornar un array (*theBooks*) que contenga los identificadores de los libros que podrían recomendarse a ese usuario. Como de costumbre, mantened el formato de definició y retorno de la función que se os proporciona y aseguraos que el código de ejemplo suministrado funciona. Si es necesario podéis definir funciones auxiliares y acceder a variables globales anteriores.

Tras la celda con el código encontraréis una celda de texto (*markdown*). En dicha celda deberéis explicar claramente los criterios que habéis seguido ara determinar los libros recomendados.

In [ ]:
# COMPLETE AND EXECUTE THE FOLLOWING CELL

def recommend_books(theUser):
    
    return recommendedBooks

# DO NOT MODIFY THE FOLLOWING LINES

theUser=19948
recommendedBooks=recommend_books(theUser)
print('The recommended books for user %d are %s '%(theUser,','.join([str(i) for i in recommendedBooks])))

**Explicad a continuación los criterios seguidos para determinar las recomendaciones**

#### **Actividad 3: Surprise**

En esta actividad trabajaréis con la biblioteca surprise. En caso de no tenerla ya instalada, ejecutad la siguiente celda.

In [ ]:
!pip install surprise

**a) (1 punto)** Haced un programa que entrene los algoritmos de Surprise SVD, KNNBasic, KNNWithMeans y NormalPredictor utilizando todos los datos de DATASET.CSV. Una vez entrenados, utilizad cda uno de los algoritmos para predecir la puntuación que daría el usuario con identificador 19948 al libro con identificador 536. Comparad esta predicción con la puntuacion real. ¿Qué algoritmo se acerca más a la puntuacón real? ¿A qué creéis que se debe? Como de costumbre, el programa que desarrolléis debe proporcionar resultados compatibles con el código de ejemplo que se os proporciona ya dentro de la siguiente celda.

In [7]:
import surprise

from pprint import pprint


real_estimate = exampleRate=dataSet.loc[(dataSet['user_id'] == 19948) & (dataSet['book_id'] == 536), 'rating'].iat[0]
closest_estimate = None

reader = surprise.dataset.Reader(sep=',', skip_lines=1)
data = surprise.dataset.Dataset.load_from_file('DATASET.CSV', reader=reader)

training = data.build_full_trainset()

method_estimate = {}
algorithms = [surprise.SVD(), surprise.KNNBasic(), surprise.KNNWithMeans(), surprise.NormalPredictor()]
for method in algorithms:
    method.fit(training)
    prediction = method.predict("19948", "536")
    result = prediction.est
    difference = abs(result - real_estimate)
    method_name = method.__class__.__name__
    method_estimate[method_name] = result
    if not closest_estimate or difference < closest_estimate[1]:
        closest_estimate = (method_name, difference)

pprint(method_estimate)
print(f"closest estimate was {closest_estimate[0]} with {closest_estimate[1]}")

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
{'KNNBasic': 4.083333333333333,
 'KNNWithMeans': 3.731020522687189,
 'NormalPredictor': 3.840953144788036,
 'SVD': 3.9583937154390134}
closest estimate was SVD with 0.041606284560986584


El algoritmo que más se ha acercado a la puntuación real es SVD (Single Value Decomposition) que nos permite hacer una descomposición de matrices de tal forma que podemos calcular sus _eigenvectors_ y así hacer predicciones sobre la misma aunque estos valores no estén definidos en el conjuntos de datos.
Sin embargo, en otras iteraciones del código se ha visto como ganador al algoritmo KNNBasic con puntuaciones muy reñinas a SVD. KNNBasic busca los k usuarios con los ratings más similares para dar una recomendación y, en este caso, como los ratings están normalizados se les puede dar más peso a estos para determinar la similitud entre usuarios.

**b) (1 punto)** Haced un programa que, para los mismos algoritmos del apartado anterior, realice una validación cruzada utilizando 5 folds y obteniendo el RMSE. ¿Qué algoritmo arroja mejores resultados? ¿Coincide con el que mejores resultados proporcionó en el apartado anterior? Justificad brevemente la respuesta. Como de costumbre, el programa debe proporcionar resultados compatibles con el código de ejemplo que se incluye dentro de la celda.

In [8]:
from surprise.model_selection import cross_validate


folds = 5
for method in algorithms:
    cross_validate(method, data, measures=['RMSE'], cv=folds, verbose=True)

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0739  1.1262  1.0850  1.0393  1.0780  1.0805  0.0278  
Fit time          0.65    0.59    0.56    0.55    0.57    0.58    0.04    
Test time         0.02    0.02    0.01    0.01    0.01    0.02    0.00    
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE of algorithm KNNBasic on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.2280  1.2683  1.2533  1.2412  1.2430  1.2468  0.0134  
Fit time          0.31    0.32    0.34    0.44    0.50    0.38    0.08    
Test ti

El algoritmo que arroja mejores resultados en la validación cruzada ha sido SVD con una media de RMSE de 1.0805 y  coincide con el algoritmo seleccionado en el apartado anterior.

#### **Actividad 4: Tendencias actuales**

**a) (1 punto)** Se puede decir que en el mundo del aprendizaje computacional ha habido un antes y un después de la popularización de las redes neuronales y el denominado *deep learning*. El campo de los recomendadores no es una excepción.

Buscad información sobre sistemas recomendadores basados en redes neuronales. Describid muy brevemente cuál es la base de su funcionamiento y qué ventajas e inconvenientes presentan respecto a métodos clásicos como los utilizados en esta PEC.

Los sistemas recomendadores tradicionales están basados en el contenido colaborativo y basan sus recomendaciones en  datos históricos para construir modelos que se ajusten a un usuario y a sus preferencias y clasificarlo correctamente junto a otros usuarios. Estos sistemas de _Machine Learning_ analizan los datos de manera lineal y requieren un conjunto menor de datos que en sistemas basados en _Deep Learning_, sin embargo, también son dependientes de un cierto volumen de estos si quieren una predicción más precisa y relevante a un nuevo usuario.
El Deep Learning es un subcampo de _Machine Learning_ que consiste en algoritmos inspirados en el funcionamiento del cerebro humano; concretamente, son familias de algoritmos que hacen una abstracción de cómo funciona una red neuronal humana. Los sistemas recomendadores basados en _Deep Learning_ permiten el procesamiento de datos de manera no lineal, son más efectivos en la minería de datos valiosos y pueden dar con características especiales en los conjuntos de datos de manera automática, sin embargo, estos modelos necesitan altas prestaciones ya que requieren de grandes volúmenes de datos para su entrenamiento y las características encontradas son difíciles de predecir o de justificar ya que se encuentran en un modelo de caja negra.

**Referencias**
* [The rise of deep learning recommender systems](https://www.dynamicyield.com/lesson/deep-learning-recommendations/)  
* [Deep Learning Based Recommender Systems](https://medium.com/sciforce/deep-learning-based-recommender-systems-b61a5ddd5456)

### Nota: Propiedad intelectual

A menudo es inevitable, al producir una obra multimedia, hacer uso de recursos creados por terceras personas. Es por tanto comprensible hacerlo en el marco de una práctica de los estudios del Grado Multimedia, siempre y cuando esto se documente claramente y no suponga plagio en la práctica.

Por lo tanto, al presentar una práctica que haga uso de recursos ajenos, se presentará junto con ella un documento en el que se detallen todos ellos, especificando el nombre de cada recurso, su autor, el lugar donde se obtuvo y el su estatus legal: si la obra está protegida por copyright o se acoge a alguna otra licencia de uso (Creative Commons, GNU, GPL ...). El estudiante deberá asegurarse de que la licencia que sea no impide específicamente su uso en el marco de la práctica. En caso de no encontrar la información correspondiente deberá asumir que la obra está protegida por copyright.

Deberán, además, adjuntar los archivos originales cuando las obras utilizadas sean digitales, y su código fuente si corresponde.

Otro punto a considerar es que cualquier práctica que haga uso de recursos protegidos por copyright no podrá en ningún caso publicarse en Mosaic, la revista del Grado en Multimedia en la UOC, a no ser que los propietarios de los derechos intelectuales den su autorización explícita.
